RFCs/rfc-0004-fast-transactions.md

```diff
@@ -0,0 +1,415 @@
+
+## Specification
+
+### Blockspace
+
+Blockspace is an ephermal L2 blockchain run inside the L1 BlockSpace node with a lifetime span comparable to the time between L1 blocks. At the end of its lifetime, this ephermal L2 blockchain transfers all the state to the L1 and disappears to be born again when needed by zkApp.
+
+![L2 drawing](https://docs.minanft.io/rfc4/teddy.png)
+
+From the [Robert Habermeier blog](https://www.rob.tech/blog/polkadot-blockspace-over-blockchains/) :
+
+“Blockspace is an ephemeral good. When you intend to commit an operation to a chain you need blockspace in-the-moment: not yesterday’s, not tomorrow’s. Blockspace is either utilized or it is not.
+
+One parallel for thinking about the design of blockspace allocation products is the cloud computing market. Cloud computing business models often have two key features: reserved instances and spot instances. Reserved instances are cheaper but guaranteed for a prolonged period of time. Spot instances are more costly, available on-demand, and ephemeral.
+
+Furthermore, by changing our perspective from blockchain-centric to blockspace-centric it becomes clear that there is no reason a blockchain or state machine should run forever. **Ephemeral blockchains are an interesting use-case I believe is highly under-explored - longer running processes should be able to offload their computations or protocols to short-lived chains just like programs running on a PC can offload computations or work to background threads**.
+
+In my opinion, the blockchain ecosystem has been thinking too small about the multi-chain world. Blockchains that start and run indefinitely with a steady pulse are an evidently inefficient mechanism. **The multi-chain of tomorrow consists of blockchains that scale and shrink on demand. It contains ephemeral chains spawned by on-chain factories, spun up by contracts and imbued with automated purpose - to complete their work a few hours later and disappear.**
+
+Our goal in Web3 is not to maximize the number of blockchains. Maximizing the number of blockchains is something I would state explicitly as a non-goal, as it primarily benefits validator cartels seeking to extract value. Our goal in Web3 is to maximize the amount of blockspace that exists and ensure it is allocated to the state machines which need it most at any time: a constant generation and allocation of global consensus resources to those who need it the most. An enterprise without waste. In other words: the most effective blockspace producer in the world.”
+
+### BlockSpace Node
+
+BlockSpace Node is a node that, in addition to the usual services of the Mina protocol node, provides the services for running blockspace - an L2 epithermal blockchain. It is up to the node to decide whether it wants to run the blockspace service, whether it is being provided to every zkApp or just to the zkApps whitelisted by this specific node, and what the fees are for running blockspace services.
```

Comment:
I'd tend to agree with what @teddyjfpender has said above, what you describe as a "BlockSpace node" is just a variant of how an L2 can be implemented on top of the MINA L1. Furthermore the RFC is missing technical details regarding the mempool usage or "guarantees".

Regarding the guarantees for client-side zkApps, you can have these features today using Zeko - at least from what i understand from their whitepaper & discussions with the team. Of course given you want to stick to the AccountUpdate model. The incentive for the Zeko sequencer operator is pretty straightforward - they get to keep the difference between what they generate in transaction fees and the cost of settlement to the L1. 

Both Zeko and Protokit allow developers to run their existing sequencing software to address the issues you've raised in this RFC. What is missing in Zeko or Protokit in your view, that is mitigated by the design described in this RFC? I personally don't see any difference in what you described vs how Zeko & Protokit operate today, and thats the main reason why i ask ^^.

Another thing i'm not entirely sure about is why you chose to skip actions / sequence state from the RFC entirely, in my understanding thats the most powerful primitive (besides cheap settlement) the L1 offers for L2s. It allows anyone to tap into the L1 consensus and use the L1 as a mempool - commonly referred to as based sequencing. Based sequencing allows the L2s to become decentralized - unlike a centralized "Blockspace node".

To address the concerns about running your own infrastructure for an L2 as an entry barrier - you don't have to do that with Zeko, and you won't have to do that after (if) we launch a general purpose rollup with Protokit either. You have to run your own infrastructure in case you want to run your own instance of Zeko, or your own instance of Protokit for an application-specific chain. I'd argue this is actually a desirable property for further decentralisation and scaling, as we know general purpose rollups or centralized infrastructure providers (like Infura) are sub optimal for the network health.

There are some additional concerns i have e.g. as you've described the depositing of assets back and forth between the short lived L2 - you're going to be limited by the amount of AUs you can pass in a single transaction today. That means if you end up merging multiple AUs together to chain their root hash state from A --> B, you won't be able to keep chaining all the other functionality that the AU offers, unless of course you design the L2 like Zeko did. Which means re-using the L1 zkapp command circuits to maintain 1:1 functionality when recursively merging AUs into your L2 state.

Apart from all the things i've touched upon above, there's a dozen different things you have to consider when designing a "VM" for a rollup or an L2. I know i'm repeating myself - but we've covered the architecture behind Protokit in our [recent workshop](https://www.youtube.com/watch?v=hqobPb3gIRo). Proof merging, transaction processing parallelisation, state transition proving parallelisation, user-land abstractions over state transitions, those are all features available today in the existing Protokit SDK. Is there anything specific thats missing from the Protokit stack to complement this RFC? Given the modular nature of the SDK, we'd be more than happy to accommodate for any missing features you identify.